x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/event/AnalyticsEvent.java

```diff
@@ -109,23 +78,48 @@ public String toString() {
     private final BytesReference payload;
 
     private final XContentType xContentType;
+    private final InetAddress clientAddress;
 
     protected AnalyticsEvent(
         String eventCollectionName,
         long eventTime,
         Type eventType,
         XContentType xContentType,
-        BytesReference payload
+        BytesReference payload,
+        @Nullable String userAgent,
+        @Nullable InetAddress clientAddress
```

Comment:
We do not need to add those field since they are supposed to be part of the payload (added by the session field parser).

Reply:
Could the user then override these as part of the session payload?

Reply:
Nope because the code is using the context to read this information (and not the payload):

```java
    private static final ObjectParser<MapBuilder<String, String>, AnalyticsEvent.Context> PARSER = ObjectParser.fromBuilder(
        SESSION_FIELD.getPreferredName(),
        (c) -> {
            MapBuilder<String, String> mapBuilder = MapBuilder.newMapBuilder();

            if (Strings.isNullOrBlank(c.clientAddress()) == false) {
                mapBuilder.put(CLIENT_ADDRESS_FIELD.getPreferredName(), c.clientAddress());
            }

            if (Strings.isNullOrBlank(c.userAgent()) == false) {
                mapBuilder.put(USER_AGENT_FIELD.getPreferredName(), c.userAgent());
            }

            return mapBuilder;
        }
    );

    static {
        PARSER.declareString(
            (b, s) -> b.put(SESSION_ID_FIELD.getPreferredName(), requireNonBlank(s, "field [id] can't be blank")),
            SESSION_ID_FIELD
        );

        PARSER.declareRequiredFieldSet(SESSION_ID_FIELD.getPreferredName());
    }

Reply:
(And the context is built from the REST request)

Reply:
that is, something like:

```java
    private static final ObjectParser<MapBuilder<String, String>, AnalyticsEvent.Context> PARSER = new ObjectParser<>(
        SESSION_FIELD.getPreferredName(),
        MapBuilder::newMapBuilder
    );

    static {
        PARSER.declareString(
            (b, s) -> b.put(SESSION_ID_FIELD.getPreferredName(), requireNonBlank(s, "field [id] can't be blank")),
            SESSION_ID_FIELD
        );

        PARSER.declareRequiredFieldSet(SESSION_ID_FIELD.getPreferredName());
    }

    private SessionAnalyticsEventField() {}

    public static Map<String, String> fromXContent(XContentParser parser, AnalyticsEvent.Context context) throws IOException {
        Map<String, String> result = PARSER.parse(parser, context).map();
        result.put(CLIENT_ADDRESS_FIELD.getPreferredName(), context.clientAddress());
        result.put(USER_AGENT_FIELD.getPreferredName(), context.userAgent());
        return Map.copyOf(result);
    }
```